relay-sampling/src/lib.rs

```diff
@@ -809,79 +839,184 @@ impl Default for SamplingMode {
-    /// context.
-    ///
-    /// This is a function separate from `get_matching_event_rule` because trace rules can
-    /// (theoretically) be applied even if there's no event. Also we expect that trace rules are
-    /// executed before event rules.
-    pub fn get_matching_trace_rule(
-        &self,
-        sampling_context: &DynamicSamplingContext,
-        ip_addr: Option<IpAddr>,
-        now: DateTime<Utc>,
-    ) -> Option<ActiveRule> {
-        self.rules.iter().find_map(|rule| {
-            if rule.ty != RuleType::Trace {
-                return None;
+    /// Generates the seed used for the random number generation that is used for sampling decisions.
+    fn get_seed(
+        event: &Event,
+        dsc: Option<&DynamicSamplingContext>,
+        has_matched_trace_rule: bool,
+    ) -> Option<Uuid> {
+        // If we are in a situation in which we have matched a trace rule, we want to use the trace id
+        // as the seed for the random number generation, to guarantee the same sampling result
+        // across transactions of the same trace. The problem here is that this logic won't help
+        // in cases in which we have multiple matches of transaction and trace rules because they will
+        // effectively apply on different payloads. The transaction rule will match the event itself which differs
+        // between each component of the trace and the trace rule will match the dynamic sampling context that
+        // is the same across all components of the trace.
+        //
+        // An example of the aforementioned case:
+        // /hello -> /world -> /transaction belong to trace_id = abc
+        // * /hello has uniform rule with 0.2 sample rate which will match all the transactions of the trace
+        // * each project has a single transaction rule with different factors (2, 3, 4)
+        //
+        // 1. /hello is matched with a transaction rule with a factor of 2 and uses as seed abc -> 0.2 * 2 = 0.4 sample rate
+        // 2. /world is matched with a transaction rule with a factor of 3 and uses as seed abc -> 0.2 * 3 = 0.6 sample rate
+        // 3. /transaction is matched with a transaction rule with a factor of 4 and uses as seed abc -> 0.2 * 4 = 0.8 sample rate
+        //
+        // We can see that we have 3 different samples rates but given the same seed, the random number generated will be the same.
+        let event_id = event.id.value();
+
+        if has_matched_trace_rule {
+            if let Some(dsc) = dsc {
+                return Some(dsc.trace_id);
             }
+        }
 
-            if !rule.condition.matches(sampling_context, ip_addr) {
-                return None;
-            }
+        if let Some(event_id) = event_id {
+            return Some(event_id.0);
+        }
 
-            rule.is_active(now)
-        })
+        None
+    }
+
+    pub fn has_unsupported_rules(&self) -> bool {
```

Comment:
Can we move this function back to the top? Just to make the diff easier to read.